lib/rucio/client/s3client.py

```diff
@@ -0,0 +1,138 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import boto3
+import os
+import logging
+from dogpile.cache.api import NO_VALUE
+
+from rucio.common.config import get_s3_credentials
+from rucio.common.cache import make_region_memcached
+from botocore.exceptions import ClientError
+from rucio.client.client import Client
+
+
+SUCCESS = 0
+FAILURE = 1
+
+REGION = make_region_memcached(expiration_time=900)
+
+
+class S3Client:
+    """S3 client class"""
+
+    S3_BASEURL = "s3"
+
+    def __init__(self, _client=None, logger=None, config: dict = None):
+        # TODO: convert dict config to S3Credentials object?
+        """
+        Initialises the basic settings for an S3Client object
+
+        :param _client:     - Optional:  rucio.client.client.Client object. If None, a new object will be created.
+        :param logger:      - Optional:  logging.Logger object. If None, default logger will be used.
+        :param config:      - Optional:  dict object with S3 credentials. If None, credentials will be loaded from default path.
+
+        """
+
+        if not logger:
+            self.logger = logging.log
+        else:
+            self.logger = logger.log
+
+        self.client = _client if _client else Client()
+        account = self.client.account
+        cred = REGION.get("s3client-%s" % account)
+        if cred is NO_VALUE:
+            if config:
+                cred = config
+            else:
+                cred = get_s3_credentials()
+            REGION.set("s3client-%s" % account, cred)
+        try:
+            self.s3 = boto3.client("s3", **cred, verify=False)
+        except Exception as error:
+            self.logger(logging.ERROR, error)
+            raise error
+
+    def bucket_create(self, path):
+        """Create an S3 bucket.
+
+        param bucket_name: Bucket path, e.g. user.dquijote:/mybucket/
+        :return: True if bucket created, else False
+        """
+        # TODO: IAM policy structure
+        bucket_name, bucket_path = path.split(":")
+
+        try:
+            self.s3.head_bucket(Bucket=bucket_name)
+        except ClientError as error:
+            # TODO: catch 409 BucketAlreadyOwnedByYou and BucketAlreadyExists (bucket can be owned by another user)
+            if error.response["Error"]["Code"] == "404":
+                self.logger(logging.DEBUG, "Creating bucket %s" % bucket_name)
+                self.s3.create_bucket(Bucket=bucket_name)
+            elif error.response:
+                self.logger(logging.ERROR, error.response["Error"]["Message"])
+                return FAILURE
+
+        try:
+            self.s3.put_object(Bucket=bucket_name, Body="", Key=bucket_path)
```

Comment:
are the parameters really uppercase? first time i see something like this for non-class objects

Reply:
according to the docs, I'm also confused about this https://boto3.amazonaws.com/v1/documentation/api /latest/reference/services/s3/client/put_object.html 🤔